No-op ConcurrentStack.PushRange(T[]) if empty

martincostello opened the pull request:
Resolves #62121.

martincostello committed:
No-op ConcurrentStack.PushRange(T[]) if empty

Resolves #62121.

martincostello commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs:
```diff
@@ -349,7 +354,6 @@ public void PushRange(T[] items, int startIndex, int count)
             if (count == 0)
                 return;
```
In retrospect, I guess this condition should be removed because it's redundant?

martincostello committed:
Add more bounds checks

Assert that ArgumentOutOfRangeException is thrown if specifying indexes for an empty array with PushRange(T[], int, int).

stephentoub commented:
Thanks.

This would address the case of `PushRange(empty)`, but not of `PushRange(empty, 0, 0)`, which should be functionally identical.

I suggest the right fix is simply to change this condition:
https://github.com/dotnet/runtime/blob/a4cab7d3206027566dc14672669caeac56ef8400/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs#L413
from
```C#
if (startIndex >= length || startIndex < 0)
```
to
```C#
if (startIndex > length || startIndex < 0)
```
or even:
```C#
if ((uint)startIndex >= length)
```
i.e. it should be fine to have the start index be at the end of the array, as long as the count is 0, which the subsequent check validates.

eiriktsarpalis commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs:
```diff
@@ -316,6 +316,11 @@ public void PushRange(T[] items)
             }
+
+            // No op if the length is zero
+            if (items.Length == 0)
```
If we're making this change I'd argue we should also do the same for `PushRange(array, 0, 0)` calls. That would also avoid the need for making this check twice.

martincostello commented:
Thanks both - I did think about that, but then asking to start from index 0 (which doesn't exist) seemed to be an equally valid mistake to me at first sight on an empty array when the range is specified, so went with this initially to gauge thoughts.

I'll update the PR as suggested 👍

martincostello committed:
Review feedback

Allow out-of-bounds index on empty array if count is zero.

stephentoub commented:
> asking to start from index 0 (which doesn't exist) seemed to be an equally valid mistake to me

This is how things generally work in .NET, e.g. you can insert at the end of a collection with the index == length,  you can take a 0-length substring at the end of a string,  etc.

stephentoub commented on src/libraries/System.Collections.Concurrent/tests/ConcurrentStackTests.cs:
```diff
@@ -63,6 +63,26 @@ public void PushRange_NoItems_NothingAdded()
+
+            s.PushRange(new int[0]);
+            Assert.True(s.IsEmpty);
+        }
```
We should also validate with a zero count at the end of a non-empty array,  e.g.
```C#
int[] arr = new int[2];
s.PushRange(arr, arr.Length, 0);
```

Clockwork-Muse commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs:
```diff
@@ -410,7 +410,7 @@ private static void ValidatePushPopRangeInput(T[] items, int startIndex, int cou
                 throw new ArgumentOutOfRangeException(nameof(count), SR.ConcurrentStack_PushPopRange_CountOutOfRange);
             }
             int length = items.Length;
-            if (startIndex >= length || startIndex < 0)
```
Doesn't this change the exception observed when `startIndex == length`, with a non-zero `count`?  You've added a test that catches `ArgumentException` for this case - is that the exception that occurs before this change?  (Also, it may be good to add a similar case for a non-empty array regardless)

Not that anybody should be catching `ArgumentException`, so as breaking changes go this is relatively minor...

martincostello committed:
Add extra test case

Add a further test case for a non-empty array but with a count parameter value of zero.
